### PR TITLE
[nrfx] Add the on-chip flash

### PR DESCRIPTION
**Pull request by supperthomas**
## 拉取/合并请求描述：(PR description)

[
on-chip flash 
test on fal read fal write fal erase
]

以下的内容不应该在提交PR时的message修改，修改下述message，PR会被直接关闭。请在提交PR后，浏览器查看PR并对以下检查项逐项check，没问题后逐条在页面上打钩。
The following content must not be changed in submitted PR message. Otherwise, the PR will be closed immediately. After submitted PR, please use web browser to visit PR, and check items one by one, and ticked them if no problem.

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [x] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other style
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 本拉取/合并请求代码是高质量的 Code in this PR is of high quality

### CODE REVIEW

File: bsp/nrf5x/libraries/drivers/drv_flash.c [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+/**
+  * @brief  Gets the page of a given address
+  * @param  Addr: Address of the FLASH Memory
+  * @retval The page of a given address
+  */
+static uint32_t GetPage(uint32_t Addr)
+{
+    uint32_t page = 0;
+	  if(Addr < (MCU_FLASH_START_ADRESS + MCU_FLASH_SIZE))
+		{
```
Comment (BernardXiong) [inline bsp/nrf5x/libraries/drivers/drv_flash.c:37]: Please take care of indentation.

File: bsp/nrf5x/libraries/drivers/drv_flash.c [hunk 2]
```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2006-2018, RT-Thread Development Team
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Change Logs:
+ * Date           Author       Notes
+ * 2018-12-5      SummerGift   first version
```
Comment (BernardXiong) [inline bsp/nrf5x/libraries/drivers/drv_flash.c:8]: Please use the correct author information.

File: bsp/nrf5x/libraries/drivers/drv_flash.c [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+ * @param addr flash address
+ * @param buf the write data buffer
+ * @param size write bytes size
+ *
+ * @return result
+ */
+
+int mcu_flash_write(rt_uint32_t addr, const uint8_t *buf, size_t size)
+{
+
```
Comment (BernardXiong) [inline bsp/nrf5x/libraries/drivers/drv_flash.c:91]: Please remove the empty line.

File: bsp/nrf5x/libraries/drivers/drv_flash.c [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        return -RT_EINVAL;
+    }
+		
+
+    if(addr % 4 != 0)
+    {
+        LOG_W("write addr should be 4-byte alignment");
+        //4byte write 
+			  //else byts
+			  nrfx_nvmc_bytes_write(addr, buf, size);
```
Comment (BernardXiong) [inline bsp/nrf5x/libraries/drivers/drv_flash.c:108]: the indentation of the code...

File: bsp/nrf5x/libraries/drivers/drv_flash.h [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+ * Date           Author       Notes
+ * 2018-12-5      SummerGift   first version
+ */
+
+#ifndef __DRV_FLASH_H__
+#define __DRV_FLASH_H__
+
+#include <rtthread.h>
+#include "rtdevice.h"
+#include <rthw.h>
```
Comment (BernardXiong) [inline bsp/nrf5x/libraries/drivers/drv_flash.h:16]: Normally, to just `include <rtdevice.h>` is enough.

File: bsp/nrf5x/libraries/drivers/drv_log.h [hunk 6]
```diff
@@ -0,0 +1,27 @@
+/*
```
Comment (BernardXiong) [inline bsp/nrf5x/libraries/drivers/drv_log.h:1]: Not recommend to use a unify header for log. You can define the log tag in each module.

File: bsp/nrf5x/nrf52840/board/board.h [hunk 7]
```diff
@@ -5,6 +5,16 @@
 
 #include "nrf.h"
 
+#define MCU_FLASH_PAGE_SIZE 0x1000
+
+#define MCU_FLASH_START_ADRESS       ((uint32_t)0x00000000)
```
Comment (BernardXiong) [inline bsp/nrf5x/nrf52840/board/board.h:10]: Or is it better to define the config in Kconfig?

File: bsp/nrf5x/nrf52840/board/fal_cfg.h [hunk 8]
```diff
@@ -0,0 +1,49 @@
+/*
+ * File      : fal_cfg.h
+ * This file is part of FAL (Flash Abstraction Layer) package
+ * COPYRIGHT (C) 2006 - 2018, RT-Thread Development Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
```
Comment (BernardXiong) [inline bsp/nrf5x/nrf52840/board/fal_cfg.h:6]: Please use correct copyright file header. For these files, only Apach License v2.0 codes are acceptable.

File: bsp/nrf5x/nrf52840/board/board.h (hunk 7 above)
Comment (supperthomas) [reply bsp/nrf5x/nrf52840/board/board.h:10]: I think this is not variable for config. This is const. Do you think const feature in Kconfig is suitable？

File: bsp/nrf5x/libraries/drivers/drv_flash.c [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+        LOG_E("ERROR: erase outrange flash size! addr is (0x%p)\n", (void *)(addr + size));
+        return -RT_EINVAL;
+    }
+
+    FirstPage = GetPage(addr);
+    NbOfPages = GetPage(addr + size - 1) - FirstPage + 1;
+
+    for (int i = 0; i < NbOfPages ; i++)
+    {
+			  result = nrfx_nvmc_page_erase((FirstPage+i) * MCU_FLASH_PAGE_SIZE);
```
Comment (BernardXiong) [inline bsp/nrf5x/libraries/drivers/drv_flash.c:141]: Please take care of indentation.

**Comment by supperthomas** [comment]
reversion 2

File: bsp/nrf5x/libraries/drivers/drv_flash.h (hunk 5 above)
Comment (supperthomas) [reply bsp/nrf5x/libraries/drivers/drv_flash.h:15]: ok

**Pull request merged**